quantized elu: eager mode QAT handling

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40284 graph mode: add hardswish inplace handling
* #40111 quantized elu: graph mode handling
* **#40104 quantized elu: eager mode QAT handling**
* #40103 quantized elu: eager mode static handling
* #40100 quantized elu: require observation

Summary:

Adds eager mode QAT handling for quantized ELU.

Test Plan:

```
python test/test_quantization.py TestQuantizationAwareTraining.test_activations
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075082](https://our.internmc.facebook.com/intern/diff/D22075082)

vkuzo committed:
quantized elu: eager mode QAT handling

Summary:

Adds eager mode QAT handling for quantized ELU.

Test Plan:

```
python test/test_quantization.py TestQuantizationAwareTraining.test_activations
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "quantized elu: eager mode QAT handling"


Summary:

Adds eager mode QAT handling for quantized ELU.

Test Plan:

```
python test/test_quantization.py TestQuantizationAwareTraining.test_activations
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "quantized elu: eager mode QAT handling"


Summary:

Adds eager mode QAT handling for quantized ELU.

Test Plan:

```
python test/test_quantization.py TestQuantizationAwareTraining.test_activations
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075082](https://our.internmc.facebook.com/intern/diff/D22075082)

[ghstack-poisoned]

vkuzo committed:
Update on "quantized elu: eager mode QAT handling"


Summary:

Adds eager mode QAT handling for quantized ELU.

Test Plan:

```
python test/test_quantization.py TestQuantizationAwareTraining.test_activations
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075082](https://our.internmc.facebook.com/intern/diff/D22075082)

[ghstack-poisoned]

jerryzh168 commented on torch/nn/qat/modules/activations.py:
```diff
@@ -41,3 +41,34 @@ def from_float(cls, mod, qconfig=None):
+            Args: `mod` a float module, either produced by torch.quantization utilities
+            or directly from user
+        """
+        assert type(mod) == cls._FLOAT_MODULE, ' qat.' + cls.__name__ + '.from_float only works for ' + \
```
btw we can probably have a helper function to assert this

vkuzo replied:
makes sense.  Might be good to do in a separate PR with just that change, since currently every eager mode module has this code snippet.

jerryzh168 replied:
yeah, sounds good

jerryzh168 commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -480,11 +480,13 @@ def __init__(self, qengine):
         x = self.quant(x)
         x = self.fc1(x)
         x = self.hardswish(x)
+        x = self.elu(x)
```
same for this one, it's better to add a dequant in the end

vkuzo committed:
Update on "quantized elu: eager mode QAT handling"


Summary:

Adds eager mode QAT handling for quantized ELU.

Test Plan:

```
python test/test_quantization.py TestQuantizationAwareTraining.test_activations
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075082](https://our.internmc.facebook.com/intern/diff/D22075082)

[ghstack-poisoned]

vkuzo committed:
Update on "quantized elu: eager mode QAT handling"


Summary:

Adds eager mode QAT handling for quantized ELU.

Test Plan:

```
python test/test_quantization.py TestQuantizationAwareTraining.test_activations
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075082](https://our.internmc.facebook.com/intern/diff/D22075082)

[ghstack-poisoned]

vkuzo committed:
Update on "quantized elu: eager mode QAT handling"


Summary:

Adds eager mode QAT handling for quantized ELU.

Test Plan:

```
python test/test_quantization.py TestQuantizationAwareTraining.test_activations
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075082](https://our.internmc.facebook.com/intern/diff/D22075082)

[ghstack-poisoned]

vkuzo committed:
Update on "quantized elu: eager mode QAT handling"


Summary:

Adds eager mode QAT handling for quantized ELU.

Test Plan:

```
python test/test_quantization.py TestQuantizationAwareTraining.test_activations
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075082](https://our.internmc.facebook.com/intern/diff/D22075082)

[ghstack-poisoned]

vkuzo committed:
Update on "quantized elu: eager mode QAT handling"


Summary:

Adds eager mode QAT handling for quantized ELU.

Test Plan:

```
python test/test_quantization.py TestQuantizationAwareTraining.test_activations
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075082](https://our.internmc.facebook.com/intern/diff/D22075082)

[ghstack-poisoned]
